packages/grid/x-data-grid/src/hooks/features/rows/gridRowsUtils.ts

```diff
@@ -213,7 +213,10 @@ export const getTreeNodeDescendants = (
     if (!skipAutoGeneratedRows || !isAutoGeneratedRow(tree[child])) {
       validDescendants.push(child);
     }
-    validDescendants.push(...getTreeNodeDescendants(tree, child, skipAutoGeneratedRows));
+    const childDescendants = getTreeNodeDescendants(tree, child, skipAutoGeneratedRows);
+    for (let j = 0; j < childDescendants.length; j++) {
+      validDescendants.push(childDescendants[j]);
+    }
```

Comment:
If https://dev.to/uilicious/javascript-array-push-is-945x-faster-than-array-concat-1oki is accurate, it should be OK-ish I guess:

![Screenshot 2024-01-20 at 20 41 11](https://github.com/mui/mui-x/assets/3165635/b07a2748-df49-478a-a097-94d24bce852c)

Reply:
Merging arrays in JS is terribly unsatisfying.

If we're worried about performance here, we could pass the result array in for mutation instead of returning results in a new array. A bit like an out-parameter in C:

```js
const getTreeNodeDescendants = (a, b, c, result = []) =>{
  // ...
  result.push(child)
  getTreeNodeDescendants(/* ... */, result) // return value doesn't matter here, it's going to be `result`
  return result
}
```

It avoids the recursive array allocations & copies that we're currently doing.